CRM/Core/Permission/Joomla.php

```diff
@@ -131,4 +135,74 @@ public function checkGroupRole($array) {
     return FALSE;
   }
 
+  /**
+   * @inheritDoc
+   */
+  public function upgradePermissions($permissions) {
```

Comment:
maybe it would be more clear if the array weren't called "permissions". That it is knowable by referencing somewhere else is not very helpful. 

It would read a lot better if the array were named to indicate that it is a map and not an array of permission objects. The line reads, for-each permission object in the permission array. You could even retrieve the key as self-documentation and just not use it...
foreach( array_flip($permRoleMap) as $perm => role) {...

... or would that be $perm => $group?? I haven't looked up that array yet. :-D